### ISSUE TITLE
Arm64 & aarch64 support?

### ISSUE THREAD

**Issue by ngam** [cylc/cylc-flow#4525]
I tried to `pip install -e` like here https://cylc.github.io/cylc-admin/howto/cylc-ui but it fails on my m1 mac (osx-arm64). Is this supposed to supported? If not, is it planned? Thanks! 

**Describe exactly what you would like to see in an upcoming release**

**Additional context**

**Pull requests welcome!**

**Comment by ngam** [comment]
To clarify, these instructions seem to work totally fine on arm64 with a minor error:

```
pip install cylc-flow
pip install cylc-uiserver
npm install configurable-http-proxy
pip install cylc-rose
```

And the minor error that doesn't seem to affect anything:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
dm-sonnet 2.0.0 requires absl-py>=0.7.1, which is not installed.
dm-sonnet 2.0.0 requires numpy>=1.16.3, which is not installed.
dm-sonnet 2.0.0 requires tabulate>=0.7.5, which is not installed.
dm-sonnet 2.0.0 requires wrapt>=1.11.1, which is not installed.
```

**Comment by ngam** [comment]
I've converted the downstream conda-forge build to noarch too --- should probably produce noarch builds there as well, it is more coherent and it will end up costing the CI less:

https://github.com/conda-forge/cylc-flow-feedstock/pull/31

**Comment by ngam** [comment]
How serious are you deps? https://github.com/cylc/cylc-flow/blob/dbaf7efc4ab4c3bfa83bdf24e100026a2ee5e79b/setup.py#L42

Seems a bit too restrictive no? Can they be relaxed a bit? A noarch build will likely be of no use if such strict deps are kept...

**Comment by ngam** [comment]
I relaxed your deps (completely eliminated all conditions) and things seem to work better by default resolve: https://anaconda.org/ngam/cylc-flow.

Looking at my conda list, I see:

```
aiofiles                  0.7.0
ansimarkup                1.5.0
colorama                  0.4.4
graphene                  3.0
jinja2                    3.0.3
metomi-isodatetime        1!2.0.2
protobuf                  3.16.0
psutil                    5.8.0
pyuv                      1.4.0
setuptools                59.2.0
urwid                     2.1.1
...
empy                      3.3.4
pandas                    1.3.4
```

**Comment by hjoliver** [comment]
> on my m1 mac (osx-arm64). Is this supposed to supported? If not, is it planned? Thanks!

Yes we should be supporting it, but probably have no one on the team with that hardware at the moment - so thanks for getting involved.

> Seems a bit too restrictive no? Can they be relaxed a bit? A noarch build will likely be of no use if such strict deps are kept...

We'll have a think about this and report back ...

**Comment by ngam** [comment]
@hjoliver thanks for the quick response and sorry I came out of nowhere. I am more than happy to help :)

A few notes:
- For full transparency, a senior colleague (at a US-based research institution) mentioned they had trouble with installing cylc-flow on an M1 mac and that's how I looked into this (I still have little idea what cylc-flow actually does, sorry, but will likely learn more soon as there will be active testing to adopt cylc-flow... and I was told there would be interest in helping maintain, develop, etc.)
- Your package (cylc-flow) seems to work totally fine when installed via pip despite the errors (your pypi wheels are noarch --- i.e. work on all platforms --- as far as I could tell that's why I initiated a rebuild on the conda-forge feedstock repo as noarch; however, it doesn't seem you embrace the noarch specifically anywhere... maybe it is not meant to be noarch? If then, the pypi wheels should be updated...) Anyway, the "alternative" installation via pip is likely more comprehensive than the usually more reliable conda installation
- Yes, please look into the numbered dependencies --- actually, the original problem my colleague had was to do with using an emulator (rosetta) to install cylc-flow x86_64 on the arm64 machine. Apparently, it took more than 24 hours to no avail. However, that's easily solvable (in case others ask, this is solved by using mamba as a drop-in replacement for conda, so `conda install mamba` and then `mamba install cylc-flow` etc. and it resolves quickly and all's well). I would gather the numbered dependencies were giving conda (through the emulator) a hard time. Now, when I rebuilt your package locally based on the conda-forge feedstock as noarch, it wasn't possible to resolve it (something about conflicting conditions when it comes to arm64 with pandas, etc.) and then I basically eliminated all numbered conditions (kept the deps, let conda resolve them by default) and it seemed to work just fine. However, we cannot/shouldn't adopt that approach in the actual feedstock (we will want to keep your exact deps, so cylc-flow developers should look into them and hopefully relax them)
- I will be doing my own testing (with the aforementioned colleague this week) and I will report back. However, if you'd like me to run specific tests or template code or something like that, I am more than happy to do so --- feel free to open an issue and tag me, or I can possibly give you an email address if you prefer. I cannot build following your development installation (see first thing in this issue); originally, I wanted to completely build this package natively on my machine...

**Comment by oliver-sanders** [comment]
> A noarch build will likely be of no use if such strict deps are kept...

Support for the M1 chips has so far been slow. Newer versions of software are more likely to have M1 support.

Our older dependencies may be causing problems?

Our dependencies are not as tight as they look (rationale bellow). We've not had issues with deps for X86 Linux or Mac OS.

Could you provide details of the dependencies you cannot acquire for M1?

>  Seems a bit too restrictive no? Can they be relaxed a bit?

We have
* 6 loosely pinned dependencies.
* 2 which we are temporally being cautious.
* 4 tightly pinned dependencies.
 
Here's an overview of the rationale behind them:

```bash
## loose pinning
                                          
# We rely on features of 1.0.0     
ansimarkup>=1.0.0                         
                                          
# Version 1 has not yet been released.
# We would expect version 1 to make breaking changes.
colorama>=0.4,<=1     
                                          
# Version 3 has only recently been released.
# We have not tested with Version 3 yet and would not expect it to work without changes.
graphene>=2.1,<3                          

# We need features from 5.6.0      
psutil>=5.6.0      

# We rely on features from 49      
setuptools>=49                            
                                          
# Version 3 has not yet been released, we would expect breaking changes when it is.
urwid==2.* 


## cautions pinning

# This project is in its early stages.
# We should expect more breaking changes until 1.0.0    
aiofiles==0.7.*

# This is a slow moving project, version 2 / 1.5 has not yet been released
# We would expect breaking changes from the next release.
pyuv==1.4.*


## tight pinning

# Jinja2 has historically made many breaking changes in minor releases.    
# These have caused a lot of user pain, we pin the Jinja2 version    
# so users can handle Jinja2 changes with Cylc version changes.
# (especially important when sharing workflows between different institutions, before this they often broke).
# Jinja2 version 3 has now been released, I think the big changes are out of the way now and the
# Jinja2 interface is more stable, we can look at relaxing this.
jinja2==2.11.*       
                                          
# This is one of our own projects, we release it alongside cylc-flow.
# We have made breaking changes, hence the current range.
# This is highly unlikely to cause conflicts.
metomi-isodatetime>=1!2.0.2, <1!2.1.0     
                                          
# This is used in networking - our schema needs to be recompiled for version changes.
# As a distributed system we need to ensure interoperability of installations on different platforms.
protobuf==3.15.*    
                                                                 
# Used for networking              
# we should be able to relax this to 19.*    
pyzmq==19.0.*     
```

I think we can safely loosen this dependency:

* `pyzmq 19.0.* => 19.*` (requires investigation into ZMQ interoperability)

With a bit of work we can bump the following deps to the latest versions (which are more likely to have aarch support):

* `pyzmq 19.0.* => 22.*` 
* `jinja2 2.11.* => 3.*` (requires investigation)
* `graphene 2.* => 3.*`
* `metomi-isodatetime 2.0.2, 2.1.0 => >=2.2,<3`

**Comment by oliver-sanders** [comment]
> (I still have little idea what cylc-flow actually does, sorry, but will likely learn more soon as there will be active testing to adopt cylc-flow... and I was told there would be interest in helping maintain, develop, etc.)

Thanks for getting involved, greatly appreciated. Also feel free to drop in on the [developer chat](https://matrix.to/#/#cylc-general:matrix.org).

> (in case others ask, this is solved by using mamba as a drop-in replacement for conda, so conda install mamba and then mamba install cylc-flow etc. and it resolves quickly and all's well).

Yes, I've converted to Mamba, Conda seems to choke on simple things sometimes.

> I will be doing my own testing (with the aforementioned colleague this week) and I will report back. However, if you'd like me to run specific tests or template code or something like that, I am more than happy to do so 

Great, thanks.

We currently have access to X86 Linux & X86 Mac OS (Catilina & Big Sur). We may be able to get access access to aarch Linux, but we don't have access to any M1 hardware at the moment.

> (something about conflicting conditions when it comes to arm64 with pandas, etc.)

Pandas is an optional dependency of `cylc report-timings` which is pinned at an older version. Would be good to upgrade this script to use a more recent version. We haven't had the time to do this yet, however, we are expecting this script to become obsolete in the near future by moving this functionality into the `cylc-ui` project.

**Comment by oliver-sanders** [comment]
Closed (for cylc-flow) by https://github.com/conda-forge/cylc-flow-feedstock/pull/33

tldr; We used the psudo dependency `__unix` to allow us to go `noarch` whilst saying Unix-only.

Will replace this with follow-up issues on cylc-uiserver and cylc-rose.
